### PR TITLE
Update proxy templates to support resource filtering

### PR DESCRIPTION
**Pull request by rosstimothy**
Historically proxy templates have only supported an exact hostname match via the `Host` field defined. This limits some of the benefits of using proxy templates and forces users to do a bunch of extra work in their ssh config. For example, if a user today wanted to leverage proxy templates to connect to a host via a label, predicate expression, or fuzzy match, they would be required to use tsh ls, find the matching UUID or hostname and provide that to tsh for evaluation with the template expression.

SSH Config:
```
Match Host *.foo.example.com
    ProxyCommand tsh proxy ssh  %r@$(tsh ls --search=$(echo %h | cut -d. -f1) -f json | jq -r '.[0].metadata.name'):%p
```

The above requires multiple sub-shells, a connection to the Auth server, and jq to be installed on the client machine in order to work.

To reduce the amount of effort required to connect to a resource proxy templates have been updated with a `Search` and a `Query` field to support fuzzy search and predicate expressions much like `tsh ls --search` and `tsh ls --query`. The order of operations for matching the target host is: 1) Host, 2) Predicate Expression, 3) Query.

In addition to the proxy template changes, tsh proxy ssh has also been updated to allow resolving the target via searching or querying -- though ONLY via proxy templates. There are no additional `--query` or `--search` flags added to tsh proxy ssh for the time being. This is mostly for CLI UX reasons since neither tsh ssh or tsh proxy ssh support those flags today, and only allow filtering by label via supplying key value pairs in the hostname  (tsh ssh alice@foo=bar,env=dev).

tsh proxy ssh is however still limited to connecting to a single host at a time. If there are no matches, or multiple matches found as a result of target resolution, then an error is returned.

With the new functionality, the above example can be simplified to the following.

SSH Config:
```
Match Host *.foo.example.com
    ProxyCommand tsh proxy ssh  %r@%h
```

Proxy Template:
```
- template: '^(.*).example.com:\d+$'
  search: "$1"
```

This shifts the responsibility of host resolution from the caller to tsh. It allows resolution to happen in the same process that will initiate the connection, reducing a dial to the cluster, and eliminating the need for any sub-shells, or extra tools outside of tsh to be present.

Changelog: Extend proxy templates to allow the target host to be resolved via a predicate expression or fuzzy matching.

### CODE REVIEW

File: docs/pages/connect-your-client/tsh.mdx [hunk 1]
```diff
@@ -1090,8 +1090,23 @@ proxy_templates:
...
   #
   # Port defaults to 3022 if not explicitly provided with `--port`.
+  # If provided, it will take precedence over host resolution via
+  # query or search.
   host: "$1:$3"
 
+  # Optional predicate expression to resolve the target host with.
+  #
+  # Query by predicate expression similar to tsh ls --query.
+  # Supports ==, !=, &&, and || (e.g. --query='labels["key1"] == "value1" && labels["key2"] != "value2"').
```
[the remark this answers is not in the document]
Comment (strideynet) [reply docs/pages/connect-your-client/tsh.mdx:1100]: Yup - predicate docs definitely more helpful!

File: tool/tsh/common/proxy.go [hunk 2]
```diff
@@ -62,13 +63,33 @@ func onProxyCommandSSH(cf *CLIConf) error {
 			return trace.Wrap(err)
 		}
 
-		targetHost, targetPort, err := net.SplitHostPort(tc.Host)
-		if err != nil {
-			targetHost = tc.Host
-			targetPort = strconv.Itoa(tc.HostPort)
+		var target string
+		switch {
```
[the remark this answers is not in the document]
Comment (strideynet) [reply tool/tsh/common/proxy.go:67]: Hmm yeah - I find with these sort of things that no matter how much you document it, you still end up with support load from people not realising one of their flags is being ignored. Perhaps better to explicitly reject, or emit a warning message that some provided configuration is being ignored due to both being specified.

**Review by strideynet — APPROVED** [review]

File: tool/tsh/common/proxy.go [hunk 3]
```diff
@@ -62,13 +63,33 @@ func onProxyCommandSSH(cf *CLIConf) error {
...
+			targetHost = cleanTargetHost(targetHost, tc.WebProxyHost(), clt.ClusterName())
+			target = net.JoinHostPort(targetHost, targetPort)
+		case len(tc.SearchKeywords) != 0 || tc.PredicateExpression != "":
+			nodes, err := client.GetAllResources[types.Server](cf.Context, clt.AuthClient, tc.ResourceFilter(types.KindNode))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			if len(nodes) == 0 {
+				return trace.NotFound("no matching targets found")
```
Comment (r0mant) [inline tool/tsh/common/proxy.go:83]: Nit: "target" seems like an internal term, call it SSH node? Something like:
```suggestion
				return trace.NotFound("no SSH nodes matching the proxy templates query expression found")
```

File: tool/tsh/common/proxy.go [hunk 4]
```diff
@@ -62,13 +63,33 @@ func onProxyCommandSSH(cf *CLIConf) error {
...
+			}
+
+			if len(nodes) > 1 {
+				return trace.BadParameter("found multiple matching targets")
```
Comment (r0mant) [inline tool/tsh/common/proxy.go:87]: Nit: Include a list of matching ones?
```suggestion
				return trace.BadParameter("found multiple SSH nodes matching proxy templates query expression: %v", xxx)
```

File: tool/tsh/common/proxy.go [hunk 5]
```diff
@@ -62,13 +63,33 @@ func onProxyCommandSSH(cf *CLIConf) error {
...
+			}
+			target = fmt.Sprintf("%s:0", nodes[0].GetName())
```
Comment (r0mant) [inline tool/tsh/common/proxy.go:89]: Nit: Add a comment on why we're specifying `:0` port.

**Review by r0mant — APPROVED** [review]

**Pull request merged**